File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.cpp
#include <string>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm> // min_element

using namespace std;

vector<pair<int, int>> dyx = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}; // RDLU

int solution(vector<vector<int>> board) {
    int answer = 0;
    int n = board.size(); 
    
    // (0, 0) -> (N-1, N-1) 까지 BFS. 각 칸의 최소 비용 cost에 적재!
    int c[n][n][4]; 
    queue< pair< pair<int, int>, pair<int, int> > > q; // {{y, x}, {dir, cost}} 
    
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            for(int k = 0; k < 4; k++){
                c[i][j][k] = INT_MAX; 
            }
        }
    }
    q.push({{0, 0}, {1, 0}}); q.push({{0, 0}, {0, 0}}); 
    c[0][0][0] = 0; c[0][0][1] = 0; c[0][0][2] = 0; c[0][0][3] = 0; 
    
    while(!q.empty()){
        auto f = q.front(); q.pop(); 
        int y = f.first.first; int x = f.first.second; 
        int dir = f.second.first; int cost = f.second.second; 
        
        for(int d = 0; d < 4; d++){
            int ny = y + dyx[d].first; 
            int nx = x + dyx[d].second; 
            if(ny >= n || nx >= n || ny < 0 || nx < 0) continue; 
            if(board[ny][nx] == 1) continue; // 벽
            if((dir+2)%4 == d) continue; // 거꾸로
            if(dir == d && c[ny][nx][d] > (cost + 100)){ // 직전
                c[ny][nx][d] = cost + 100; 
                q.push({{ny, nx}, {d, cost+100}}); 
            } else if(dir != d && c[ny][nx][d] > (cost + 600)){ // 꺽기
                c[ny][nx][d] = cost + 600; 
                q.push({{ny, nx}, {d, cost+600}});
            }
        }
    }
    
    // 경주로를 건설하는 데 필요한 최소 비용
    return answer = *min_element(c[n-1][n-1], c[n-1][n-1]+4);
}

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.*; 

class Data{
    int dir,  cost,  y,  x; 
    public Data(int dir, int cost, int y, int x){
        this.dir = dir; 
        this.cost = cost; 
        this.y = y; 
        this.x = x; 
    }
}

class Solution {
    
    int[][] dyx = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}; 
    Queue<Data> q = new LinkedList<>(); 
    
    // 경주로 건설
    public int solution(int[][] board) {
        int n = board.length; 
        int answer = 0;
        
        // 들어온 방향 고려! 비용 저장
        int[][][] c = new int[n][n][4]; // 기본으로 0으로 초기화
        for(int y = 0; y < n; y++){ // 초기화
            for(int x = 0; x < n; x++){
                for(int d = 0; d < 4; d++)
                    c[y][x][d] = Integer.MAX_VALUE; 
            }
        }
        
        // (0, 0)에서 bfs로 (n-1, n-1)로 이동
        q.offer(new Data(0, 0, 0, 0)); c[0][0][0] = 0; 
        q.offer(new Data(1, 0, 0, 0)); c[0][0][1] = 0; 
        while(!q.isEmpty()){
            Data t = q.peek(); q.poll(); 
            int dir = t.dir;
            int cost,  y,  x; 
            
            for(int d = 0; d < 4; d++){ // 진행 방향
                if(d == (t.dir + 2)%4) continue; // 되돌아가는 선택지 제거
                int[] yx = dyx[d];
                int ny = t.y + yx[0]; 
                int nx = t.x + yx[1]; 
                if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue; // 경계
                if(board[ny][nx] == 1) continue; // 벽
                
                // 비용 갱신
                if(d == t.dir){ // 직선
                    if(c[ny][nx][d] > (c[t.y][t.x][t.dir] + 100)){
                        c[ny][nx][d] = c[t.y][t.x][t.dir] + 100;
                        q.offer(new Data(d, c[t.y][t.x][t.dir] + 100, ny, nx));
                    }
                }
                else{ // 코너까지
                    if(c[ny][nx][d] > (c[t.y][t.x][t.dir] + 600)){
                        c[ny][nx][d] = c[t.y][t.x][t.dir] + 600;
                        q.offer(new Data(d, c[t.y][t.x][t.dir] + 600, ny, nx));
                    }
                }
                
            }
            
        }
        
        // 견적서 작성을 위해 경주로를 건설하는 데 필요한 최소 비용
        return answer = Arrays.stream(c[n-1][n-1]).min().getAsInt();
    }
}